### PR TITLE
Sines 1.0.0

### PR DESCRIPTION
**Pull request by aidanreilly**
* Added optional z_tuning support
* Updated the UI
* Created a nice init patch
* Added a per-env delay param. The delay fires at the end of the env cycle
* Added global and per-env delay random multiplier params to add variation to the cycling envelopes
* Changed the norns control scheme E1 to scroll through each line of GUI params, E2/E3 to change values. No more button + knob combos
* Added configurable crow options: 4 * 1v/o, or 2 * v/o + env
* Added a new pseudo "envelope-follower" control option
* removed the tanh limiter in favour of a more basic/clean limiter

### CODE REVIEW

**Comment by p3r7** [comment]
maybe the current tuning and/or the hz value of selected sine, idk.

the UI is already pretty crammed, isn’t it?

**Comment by aireilly** [comment]
Yes current tuning be it 12t or utonal. And hz of current sine I guess? 🤷‍♀️

**Comment by aireilly** [comment]
Maybe just current value in hz squished in somewhere.

**Comment by aireilly** [comment]
I think we can lose the master vol + panning info from the main UI. panning can be relegated to the settings menu, and since we shouldn't need to worry about clipping now (vol * 0.2 change), we don't need the master vol control either

**Comment by p3r7** [comment]
w/ the inclusion of the tahn soft limiting, that'd be one more reason to kill the volume control & display.

(also https://github.com/aidanreilly/sines/pull/16 was only necessary due to this one volume param)

**Comment by aireilly** [comment]
Yes indeed! Hoping to get to this in January.

**Comment by aidanreilly** [comment]
Did a big refactor - the four spaces tab stops were annoying, so everything is a bit all over the place.

**Comment by aireilly** [comment]
I think also we need to make sure that these changes are wholly opt in - z_tuning params and GUI elements should not be present unless z_tuning is explicitly enabled.

**Comment by aidanreilly** [comment]
currently, if the z_tuning mod is not enabled, a fatal error occurs on script load. Not sure how to work around this :/

```
### SCRIPT ERROR: load fail
/home/we/norns/lua/core/mods.lua:120: table index is nil
```

**Comment by aireilly** [comment]
Ah just move the initial z_tuning require inside the if z_tuning statement. Derp

**Comment by aidanreilly** [comment]
```
z_tuning = require 'z_tuning/lib/mod'
```
you can't set this inside the if statement that looks for z_tuning 🤷

**Comment by aireilly** [comment]
Or just require z_tuning as a prerequisite

**Comment by p3r7** [comment]
maybe do something like?

```lua
local ztuning 
if util.file_exists(_path.code..'z_tuning/lib/mod.lua') then
  ztuning = require('z_tuning/lib/mod')
end
```

EDIT: fixed missing lua file extension inside call to `util.file_exists`

File: sines.lua [hunk 1]
```diff
@@ -245,6 +290,12 @@ function set_notes()
   end
 end
 
+function set_crow_notes(chord)
+  for i = 1, 4 do
+    crow.output[i].volts = notes[crow_chords[chord][i]]
```
Comment (aidanreilly) [inline sines.lua:295]: (note_num-60)/12

File: sines.lua [hunk 2]
```diff
@@ -59,45 +63,133 @@ local step = 0
...
+        while true do
+          clock.sleep(step_s)
+          if screen_dirty then
+            set_active()
+            redraw()
+            screen_dirty = false
+          end
+          if params:get("play_mode") == 1 then
+            fader_follow_vals[i] = fader_follow_vals[i] - 1
+            engine.vol(i -1 , util.linlin(0, 127, 0.0, 1.0, fader_follow_vals[i]) * 0.2)
```
Comment (aireilly) [inline sines.lua]: engine.vol(i -1 , util.linexp(0, 127, 0.0, 1.0, fader_follow_vals[i]) * 0.2)

File: sines.lua [hunk 3]
```diff
@@ -384,16 +398,32 @@ function set_crow_note_env_pairs()
...
-  crow.output[4].action = "loop( { ar("..crow_out_4_env_atk..", "..crow_out_4_env_dec..", 7, 'log') } )"
-  crow.output[4]()
+  if params:string("crow_env_delay_pos" .. i) == "before" then
+    -- _/\
+    crow.output[4].action = "loop{ to(0, dyn{crow_env_delay4 = 0.0}), to(7, dyn{crow_attack4 = 0.01}), to(0, dyn{crow_decay4 = 0.1}) }"
+  elseif params:string("crow_env_delay_pos" .. i) == "after" then
+    -- /\_
+    crow.output[4].action = "loop{ to(7, dyn{crow_attack4 = 0.01}), to(0, dyn{crow_decay4 = 0.1}), to(0, dyn{crow_env_delay4 = 0}) }"
+  end
+    crow.output[4].dyn.crow_env_delay4 = params:get("env_delay" .. i)/100
```
Comment (aireilly) [inline sines.lua]: crow.output[4].dyn.crow_env_delay4 = params:get("env_delay" .. i)/100 + math.random() * params:get("env_delay_rand" .. i))

**Pull request merged**